sdk/include/opentelemetry/sdk/common/http_client.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <chrono>
+#include <string>
+#include <vector>
+#include "opentelemetry/nostd/function_ref.h"
+#include "opentelemetry/version.h"
+
+// Usage Example
+//
+// struct ReponseHandler: public HttpResponseHandler {
+//      void OnHttpResponse(HttpResponse& res) noexcept override
+//      {
+//           if (res.IsSuccess())
+//              .. process response
+//      }
+//  };
+//
+//   HttpSessionManager sessionManager; // implementer can provide singleton implementation for it
+//   auto session = sessionManager.createHttpSession("localhost", 8000);
+//   auto request = session->CreateRequest();
+//   request->AddHeader(..);
+//   ResponseHandler res_handler;
+//   session->SendRequest(res_handler);
+//   session->FinishSession() // optionally in the end.
+//   ...shutdown
+//    sessionManager.FinishAllSessions()
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace common
+{
+namespace http
+{
+
+enum class HttpMethod
+{
+  HTTP_GET,
+  HTTP_POST,
+  HTTP_PUT,
+  HTTP_OPTIONS,
+  HTTP_HEAD,
+  HTTP_PATCH,
+  HTTP_DELETE
+};
+
+enum class HttpSessionState
+{
+  CREATED,  // session object is created
+  ONGOING,  // http(s) request is ongoing
+  FINISHED  // https(s) request is finished
+};
+
+using HttpHeader     = std::pair<std::string, std::string>;
+using HttpBody       = std::vector<uint8_t>;
+using HttpStatusCode = uint16_t;
+
+class HttpRequest
+{
+
+public:
+  virtual void setMethod(HttpMethod method) noexcept = 0;
+
+  virtual void setURI(const std::string &uri) noexcept = 0;
+
+  virtual void SetBody(HttpBody &body) noexcept = 0;
+
+  virtual void AddHeader(const HttpHeader &header) noexcept = 0;
+
+  virtual void ReplaceHeader(const HttpHeader &header) noexcept = 0;
+
+  virtual void SetTimeoutMs(std::chrono::milliseconds timeout_ms) noexcept = 0;
+};
+
+class HttpResponse
+{
+public:
+  virtual const HttpBody &GetBody() const noexcept = 0;
+
+  virtual bool GetNextHeader(nostd::function_ref<bool(HttpHeader)> callable) const noexcept = 0;
+
+  virtual bool GetNextHeader(const std::string &key,
+                             nostd::function_ref<bool(HttpHeader)> callable) const noexcept = 0;
+
+  virtual HttpStatusCode GetStatusCode() const noexcept = 0;
+
+  virtual bool IsSuccess() const noexcept = 0;
```

Comment:
`IsSuccess()` is false if there is no HttpResponse from the server. Which can be due to read/write errors, timeout, max redirect counts. So `404` and `200` are success. `GetStatusCode()` would be undefined if `IsSuccess()` is false. And the actual error message would be available using `GetErrorMessage()`.

Reply:
In case `IsSuccess` is false, are all other methods except `GetErrorMessage` undefined? Like `GetNextHeader` or `GetBody`?

Reply:
Do you think it would be more ergonomic to have an `OnError` method on the `HttpResponseHandler`? Then we could skip `IsSucces` and `GetErrorMessage` here and instead just have an `OnError` method that receives an error message.

Reply:
Nice suggestion. Have modified the API according.